paper-server/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java

```diff
@@ -801,6 +806,54 @@ public boolean dropItem(boolean dropAll) {
         // Paper end - Fix HumanEntity#drop not updating the client inv
     }
 
+    @Override
+    public @Nullable Item dropItem(final int slot, final int amount) {
+        // Make sure the slot is in bounds
+        if (slot < 0 || slot >= this.inventory.getSize()) {
+            throw new IndexOutOfBoundsException("Slot " + slot + " out of range for inventory of size " + this.inventory.getSize());
+        }
+
+        return dropItemRaw(this.inventory.getItem(slot), amount, false);
+    }
+
+    @Override
+    public @Nullable Item dropItem(final @NotNull EquipmentSlot slot, final int amount) {
+        return dropItemRaw(this.inventory.getItem(slot), amount, false);
+    }
+
+    @Override
+    public @Nullable Item dropItemRandomly(final int slot, final int amount) {
+        // Make sure the slot is in bounds
+        if (slot < 0 || slot >= this.inventory.getSize()) {
+            throw new IndexOutOfBoundsException("Slot " + slot + " out of range for inventory of size " + this.inventory.getSize());
+        }
+
+        return dropItemRaw(this.inventory.getItem(slot), amount, true);
+    }
+
+    @Override
+    public @Nullable Item dropItemRandomly(final @NotNull EquipmentSlot slot, final int amount) {
+        return dropItemRaw(this.inventory.getItem(slot), amount, true);
+    }
+
+    private @Nullable Item dropItemRaw(final ItemStack originalItemStack, final int amount, final boolean throwRandomly) {
+        if (originalItemStack == null || originalItemStack.isEmpty() || amount <= 0) {
+            return null;
+        }
+
+        final ItemStack clonedItemStack = originalItemStack.clone();
+        final int droppedAmount = Math.min(clonedItemStack.getAmount(), amount);
+        clonedItemStack.setAmount(droppedAmount);
+
+        final ItemEntity droppedEntity = this.getHandle().drop(CraftItemStack.asNMSCopy(clonedItemStack), throwRandomly, true);
+        if (droppedEntity == null) {
+            return null;
+        }
+
+        originalItemStack.setAmount(originalItemStack.getAmount() - droppedAmount);
```

Comment:
I think this logic could be made nicer by operating on the underlying nms.ItemStack. All the Inventory#getItem should be returning mirrors (lets double check to be sure), but if they are, you can just get the underlying stack with `CraftItemStack#unwrap`. Then you can use the nms `.split` method which nicely handles all of the splitting/copying for you.

Reply:
Yeah they are all returning mirrors.

Reply:
Perfect, then just unwrap the mirror and use the methods on the nms ItemStack